### PR TITLE
Use boxed slices for wheel filename tags

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

Easy trick to reduce the size of the struct, since these are never mutated.

### CODE REVIEW

File: crates/distribution-filename/src/wheel.rs [hunk 1]
```diff
@@ -92,12 +86,6 @@ impl WheelFilename {
         // The wheel filename should contain either five or six entries. If six, then the third
         // entry is the build tag. If five, then the third entry is the Python tag.
         // https://www.python.org/dev/peps/pep-0427/#file-name-convention
-        //
-        // 2023-11-08(burntsushi): It looks like the code below actually drops
-        // the build tag if one is found. According to PEP 0427, the build tag
-        // is used to break ties. This might mean that we generate identical
-        // `WheelName` values for multiple distinct wheels, but it's not clear
-        // if this is a problem in practice.
```
Comment (charliermarsh) [inline crates/distribution-filename/src/wheel.rs:100]: This has since been fixed, but I forgot to remove the TODO at the time.

**Comment by charliermarsh** [comment]
That's surprising, will take a look.

**Review by BurntSushi — APPROVED** [review]
I like it.

We could probably do more work to shrink this type if we think it's causing issues.

**Comment by charliermarsh** [comment]
Why do you think the benchmarks regressed? :(

When I just did collect and didn't pre-allocator the vector by scanning the string twice, it regressed even more.

**Comment by BurntSushi** [comment]
> Why do you think the benchmarks regressed? :(
> 
> When I just did collect and didn't pre-allocator the vector by scanning the string twice, it regressed even more.

I spent some time looking at the flamegraph diff, and nothing sticks out at me. So for this, I think my next step would be to:

1. Run the benchmarks locally by measuring wall clock time instead of instruction count, which is what I believe codspeed is doing, right? If the wall clock time has no difference, then I'd ship it.
2. If there is a difference, then I'd look to see if profiles locally say more than the codspeed flamegraphs. Failing that, I'd start looking at the actual codegen differences.

In terms of _why_ this is happening, I don't actually have a great guess. Maybe this change increases instruction counts by doing the `Vec -> Box` conversion and that in turn causes the regression? That guess is making _a lot_ of assumptions though (codspeed's cost model, that `Vec -> Box` conversion is even adding instructions and probably more) that aren't necessarily correct.

**Pull request closed without merging**